plugins/outputs/sql/sql.go

```diff
@@ -237,28 +252,86 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
 		strings.Join(placeholders, ","))
 }
 
+func (p *SQL) createTable(metric telegraf.Metric) error {
+	tablename := metric.Name()
+	stmt := p.generateCreateTable(metric)
+	if _, err := p.db.Exec(stmt); err != nil {
+		return fmt.Errorf("creating table failed: %w", err)
+	}
+	p.tables[tablename] = make(map[string]bool)
+	// modifying the table schema is opt-in
+	if p.TableUpdateTemplate == "" {
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+	return nil
+}
+
+func (p *SQL) createColumn(tablename, column, columnType string) error {
+	// update table cache with the latest table schema
+	// some servers may lack support for "IF NOT EXIST"
+	// clause in which case, we don't want to burden users
+	// with writing complex "IF NOT EXIST" workarounds
+	if !p.tables[tablename][column] {
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+	if !p.tables[tablename][column] {
+		createColumn := p.generateAddColumn(tablename, column, columnType)
+		if _, err := p.db.Exec(createColumn); err != nil {
+			return fmt.Errorf("creating column failed: %w", err)
+		}
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+	return nil
+}
+
 func (p *SQL) tableExists(tableName string) bool {
 	stmt := strings.ReplaceAll(p.TableExistsTemplate, "{TABLE}", quoteIdent(tableName))
 
 	_, err := p.db.Exec(stmt)
 	return err == nil
 }
 
+func (p *SQL) updateTableCache(tablename string) error {
+	stmt := strings.ReplaceAll(p.tableListColumnsTemplate, "{TABLE}", quoteStr(tablename))
+
+	columns, err := p.db.Query(stmt)
+	if err != nil {
+		return fmt.Errorf("fetching columns for table(%s) failed: %w", tablename, err)
+	}
+	defer columns.Close()
+
```

Comment:
wondering if table doesn't exist, does it make sense to initialise the map?
```
// Initialize the map if it doesn't exist
    if p.tables[tablename] == nil {
        p.tables[tablename] = make(map[string]bool)
    }
```

Reply:
Right now it shouldn't be an issue, but a code refactor might make it easy to assume that it's a self-contained function. I'll update the function as suggested.